rafthttp/snapshot_hub.go

```diff
@@ -0,0 +1,170 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rafthttp
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+// timeout that waits snapshot to be called to write out
+// Before snapshot is called to write out, raft snapshot message needs to be sent out,
+// then let remote member send request to ask for snapshot, so at least
+// two maximal RTT is required, which is 10s.
+const snapshotWaitTimeout = 10 * time.Second
+
+type snapshotHub struct {
+	tr         http.RoundTripper
+	urlPickers map[types.ID]*urlPicker
+	r          Raft
+	snapst     SnapshotStore
+
+	// protect fields below
+	mu sync.Mutex
+	// whether it has registered snapshot that is going to be written out or closed
+	hasSnap   bool
+	snapTo    types.ID
+	snapIndex uint64
+	// when it fires, it closes the registered snapshot
+	snapClose *time.Timer
+}
+
+func newSnapshotHub(tr http.RoundTripper, urlPickers map[types.ID]*urlPicker, r Raft, snapst SnapshotStore) *snapshotHub {
+	return &snapshotHub{
+		tr:         tr,
+		urlPickers: urlPickers,
+		r:          r,
+		snapst:     snapst,
+	}
+}
+
+// registerSnap registers snapshot in snapshot hub.
+// The registered snapshot waits to be written out through writeSnapTo func.
+// If it exceeds wait timeout, the snapshot is closed and deregistered.
+// At most one snapshot could be registered at the same time, or it panics.
+func (s *snapshotHub) registerSnap(to types.ID, index uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.hasSnap {
+		plog.Panicf("unexpected snapshot register when previous one has not finished")
+	}
+	s.hasSnap = true
+	s.snapTo = to
+	s.snapIndex = index
+	s.snapClose = time.AfterFunc(snapshotWaitTimeout, func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if !s.hasSnap {
+			return
+		}
+		s.hasSnap = false
+		// close the snapshot in the store
+		s.snapst.WriteSnapTo(&errWriter{errors.New("cancelled")})
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out due to timeout", index, s.snapTo)
+		s.r.ReportSnapshot(uint64(to), raft.SnapshotFailure)
+	})
+}
+
+// writeSnapTo writes the snapshot with the given index into the given writer.
+func (s *snapshotHub) writeSnapTo(w io.Writer, index uint64) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.hasSnap {
+		return fmt.Errorf("failed to find available snapshot")
+	}
+	if s.snapIndex != index {
+		return fmt.Errorf("failed to find wanted snapshot")
```

Comment:
can we be more specific with the error? failed to find the snapshot with index %d, index